test/http-agent.js

```diff
@@ -0,0 +1,268 @@
+        await new Promise((resolve, reject) => {
+            upstreamProxyServer = proxy(http.createServer());
+            upstreamProxyServer.listen(upstreamProxyPort, (err) => {
+                if (err) return reject(err);
+                resolve();
+            });
+        });
+
+        // Setup main proxy server with custom agents
+        mainProxyServerPort = freePorts.shift();
+    });
+
+    after(() => {
+        if (targetServer) targetServer.close();
+        if (upstreamProxyServer) upstreamProxyServer.close();
+        if (mainProxyServer) mainProxyServer.close(true);
+    });
+
+    it('accepts httpAgent and httpsAgent in prepareRequestFunction', async () => {
+        const httpAgent = new http.Agent({ keepAlive: true });
+        const httpsAgent = new https.Agent({ keepAlive: true });
+
+        let agentsPassed = false;
+
+        if (mainProxyServer) await mainProxyServer.close(true);
+
+        mainProxyServer = new Server({
+            port: mainProxyServerPort,
+            prepareRequestFunction: () => {
+                agentsPassed = true;
+                return {
+                    upstreamProxyUrl: `http://localhost:${upstreamProxyPort}`,
+                    httpAgent,
+                    httpsAgent,
+                };
+            },
+        });
+
+        await mainProxyServer.listen();
+
+        // Make HTTP request through the proxy
+        await new Promise((resolve, reject) => {
+            request({
+                url: `${targetServerUrl}/hello-world`,
+                proxy: `http://localhost:${mainProxyServerPort}`,
+            }, (error, response) => {
+                if (error) return reject(error);
+                expect(response.statusCode).to.eql(200);
+                resolve();
+            });
+        });
+
+        expect(agentsPassed).to.be.true;
+
+        // Cleanup agents
+        httpAgent.destroy();
+        httpsAgent.destroy();
+    });
+
+    it('reuses connections with keepAlive agents (sticky IP simulation)', async () => {
+        if (mainProxyServer) await mainProxyServer.close(true);
+
+        const httpAgent = new http.Agent({
+            keepAlive: true,
+            maxSockets: 1,
+        });
+
+        let requestCount = 0;
+        const socketIds = [];
+
+        // Track sockets to verify reuse
+        httpAgent.on('free', (socket) => {
+            const socketId = `${socket.remoteAddress}:${socket.remotePort}`;
+            socketIds.push(socketId);
+        });
+
+        mainProxyServer = new Server({
+            port: mainProxyServerPort,
+            prepareRequestFunction: () => {
+                requestCount++;
+                return {
+                    upstreamProxyUrl: `http://localhost:${upstreamProxyPort}`,
+                    httpAgent,
+                };
+            },
+        });
+
+        await mainProxyServer.listen();
+
+        // Make multiple requests
+        for (let i = 0; i < 3; i++) {
```

Comment:
Nice to have: would be good to actually use concurrent reqeusts here to make test more challenging